### PR TITLE
storage: Collect all software devices into one panel

### PR DESCRIPTION
**Pull request by mvollmer**
The reasons for doing this is to remove clutter from the UI for things that people don't have and don't care about.  For example, if you are not interested in VDO, the empty VDO panel is just distracting and uses space. With this, adding one more entry to the "Create" menu has little cost, and to show this, I have added "Create Loop Device", which is not something that we would want to have a dedicated panel for, I guess.

## For release notes

image

### CODE REVIEW

**Comment by mvollmer** [comment]
Looks like this:
image

I left the iSCSI Targets as a separate panel since what it shows is not a storage device, and you can't click on it to go any details.

**Comment by andreasn** [comment]
We discussed this on IRC a bit, but putting it here too, for everyone else.

The dropdown is a bit unclear, so I think this could work instead.
Screenshot from 2019-05-13 14-03-31

I would also recommend keeping block devices in it's own group.

**Comment by garrett** [comment]
FWIW: With an "Add" button next to "Devices", it looks like you're adding devices.

I'm not sure if something like "Create" would be much better though — you're not creating devices. :wink:

**Comment by mvollmer** [comment]
> I would also recommend keeping block devices in it's own group.

I tend to agree. What would the title of the new unified panel be then?  Is "Devices" vs "Drives" clear enough?

**Comment by andreasn** [comment]
>  Is "Devices" vs "Drives" clear enough?

They sounds (and look) very alike. Would "Block Devices" for the latter be clearer maybe?

**Comment by martinpitt** [comment]
Adding "needsdesign" label for now. Please unlabel and re-request my review once you are done. Thanks!

**Comment by andreasn** [comment]
I get a distorted spinner when I create a RAID device.

Screenshot from 2019-05-17 12-09-01

**Comment by andreasn** [comment]
> Would "Block Devices" for the latter be clearer maybe?

Err...scratch that. The labels _Devices_ and _Block Devices_ would of course make it even more vague. I think I'm good with _Devices_ and _Drives_ for now.

**Comment by mvollmer** [comment]
> I get a distorted spinner when I create a RAID device.

This is already broken on master.  I make sure #11825 fixes this.

**Comment by martinpitt** [comment]
Adding needswork, this breaks tests a lot. Is this still "needsdesign"? It seems "Devices and Drives" is the design decision?

**Comment by mvollmer** [comment]
Let's go with "Devices".  @andreasn, any objections?

(One option is to not have any title at all, and let people figure it out by looking at the content.)

**Comment by mvollmer** [comment]
VDO-on-demand test on rhel 8 is failing.

**Comment by andreasn** [comment]
Yeah, `Devices` works for me.

The lines tends to move around when I hit the button for some reason.
https://andreasn.portknox.net/s/KQHH77iyLbRaDJD

**Comment by garrett** [comment]
It looks like something is changing when the menu button gets focus. Perhaps padding or margin in the icon?

**Comment by garrett** [comment]
It's due to the `dropdown-toggle` class. Specifically:

```css
.bootstrap-select.btn-group .btn:focus, .content-header-extra .dropdown-toggle:focus, .dropdown-toggle:focus, .form-control:not(.ct-select):focus, .input-group-addon:focus {
    padding-bottom: 0;
    border-bottom-width: 2px;
}
```

(Padding-bottom is overridden and does not apply)

and

```css
.bootstrap-select.btn-group .btn:active, .bootstrap-select.btn-group .btn:focus, .bootstrap-select.btn-group .btn:hover, .content-header-extra .dropdown-toggle:active, .content-header-extra .dropdown-toggle:focus, .content-header-extra .dropdown-toggle:hover, .dropdown-toggle:active, .dropdown-toggle:focus, .dropdown-toggle:hover, .form-control:not(.ct-select):active, .form-control:not(.ct-select):focus, .form-control:not(.ct-select):hover, .input-group-addon:active, .input-group-addon:focus, .input-group-addon:hover {
    border-color: var(--pf-global--BorderColor--300);
    border-bottom-color: var(--pf-global--primary-color--100);
    color: var(--pf-global--Color--100);
}
```

(Color is overridden and does not apply)

Both of these come from `patternfly-overrides.less`. A `btn` isn't expected to be of type `dropdown-toggle`.

I looked into changing the CSS (by adding `:not(.btn)` to the `.dropdown-toggle`), but then there are other assumptions about `btn`s and `dropdown-toggle`s that come into play, so the border & border radius aren't set properly.

The easiest, and probably most correct solution is to remove the "dropdown-toggle" class from the HTML in the `StorageBarMenu` block — then everything works as expected.

**Comment by garrett** [comment]
As I'm not sure this is the _correct_ fix, I've sent a PR for this PR: :wink:
https://github.com/mvollmer/cockpit/pull/44

**Comment by mvollmer** [comment]
> As I'm not sure this is the _correct_ fix, I've sent a PR for this PR: wink
> [mvollmer#44](https://github.com/mvollmer/cockpit/pull/44)

Nice, thanks!

**Comment by mvollmer** [comment]
I want to have another look at the on-demand installation, and maybe use some bits from #13027.

**Comment by mvollmer** [comment]
Demo: https://www.youtube.com/watch?v=Z1uB3MWPGEM

**Comment by mvollmer** [comment]
"needsdesign" for the "Details" revealing part.

**Comment by garrett** [comment]
The "Create VDO Device" dialog feels wrong to me.

- How does error handling of VDO device creation work when VDO support is not yet installed?
- What happens when someone configures a VDO device and the packages fail to install?
- How long does it take to install VDO support? (We should, at a minimum, have a % shown.)

I think the VDO support would need to be split out as a first step before configuring a VDO device. Unfortunately, it takes a while to install the support packages for VDO, but having it inline makes the dialog doing all sorts of different things with very fragile error handling (if any at all), so I think it's "the lesser of two evils".

**Comment by mvollmer** [comment]
"needsdesign" for the "Details" revealing part.

> The "Create VDO Device" dialog feels wrong to me.

Fair enough.

>     * How does error handling of VDO device creation work when VDO support is not yet installed?

This works just like before.  You can think of this as executing "dnf install vdo && vdo create ..." on the command line.  Or as typing "vdo create ..." into the command line and then bash says "vdo: command not found" and the magic PackageKit hook asks you whether you want to install the "vdo" package or not, and if you say "yes", it will do that and then immediately execute the command that you typed.
 
>     * What happens when someone configures a VDO device and the packages fail to install?

The VDO device will not get created and the dialog has to be closed.

>     * How long does it take to install VDO support? (We should, at a minimum, have a % shown.)

It depends.  A couple of seconds in the typical case.

> I think the VDO support would need to be split out as a first step before configuring a VDO device. Unfortunately, it takes a while to install the support packages for VDO, but having it inline makes the dialog doing all sorts of different things with very fragile error handling (if any at all), so I think it's "the lesser of two evils".

What about having a section in the menu dedicated to installation?

   Create MDRAID Device
   Create Volume Group
   ------------------------
   Install VDO support

Once VDO support has been installed, the menu changes into

   Create MDRAID Device
   Create Volume Group
   Create VDO Device

**Comment by garrett** [comment]
>  You can think of this as executing "dnf install vdo && vdo create ..." on the command line.

It's more like `vdo create && dnf install vdo` from the user's viewpoint, which is a bit weird.

> The VDO device will not get created and the dialog has to be closed.

Yeah, that's what I was wondering. This feels wrong too.

So it would feel like this:
1. Choose to create VDO device
2. Invest (a little bit of) time & energy
3. Wait
4. Have package installation fail

Whereas if it is a front-loaded package install, we're more up-front with the error and not asking someone to invest time & energy on something that won't work.

> What about having a section in the menu dedicated to installation?

I thought about this, but then we're going through steps to install the software when they don't want to install software; they want to create the device.

We'd probably want to keep it as creating a VDO device, but have a first-time dialog that installs the necessary support and then immediately kicks over to the create VDO device dialog. It effectively almost becomes something like a 2-step "wizard" when there isn't VDO support.

We would probably want a flow that's something along the lines of:
1. Summary dialog
   - VDO support is not present
   - We'll install VDO support, which means X/Y/Z
   - Cancel / Install support
2. Show installation progress
3. _Then_ show the create VDO

**Comment by mvollmer** [comment]
> So it would feel like this:
> 
>     1. Choose to create VDO device
> 
>     2. Invest (a little bit of) time & energy
> 
>     3. Wait
> 
>     4. Have package installation fail

But note that the dialog will do a extensive check re whether the installation is possible immediately when it is opened.  If there is `insights-client` package, or its dependencies can not be satisfied, then the user will know immediately, and not only after pressing "Create".

The only things that might go wrong after pressing "Create" are things like network failures, disk full, and bugs in packages.  I think those might be rare enough.

> We would probably want a flow that's something along the lines of:
> 
>     1. Summary dialog
>        
>        * VDO support is not present
>        * We'll install VDO support, which means X/Y/Z
>        * Cancel / Install support
> 
>     2. Show installation progress
> 
>     3. _Then_ show the create VDO

Oh, that's what we had in an earlier (much simpler) version of this PR. :-)  I put that back.

**Comment by garrett** [comment]
I was thinking something like this would happen the first time someone goes to create a VDO disk:

vdo-install

(When VDO support exists, it should skip to the big dialog, of course.)

Checking for VDO support should be instant and done on page load. (Don't scan for the package via PackageKit when the dialog appears, in other words.)

**Comment by mvollmer** [comment]
> > If there is `insights-client` package
> 
> I think that's outside the scope of this page. wink
> 
> But, yeah, I think Insights should be handled differently, even if it also installs packages on the fly.

Whoa.  That was meant to be "If there is no `vdo` package ..."

**Comment by garrett** [comment]
> Whoa. That was meant to be "If there is no `vdo` package ..."

Yeah, I figured. :wink:

**Comment by mvollmer** [comment]
> I was thinking something like this would happen the first time someone goes to create a VDO disk:

We have an existing "Install Software" dialog that I would use for that.  That dialog looks a bit different and doesn't have progress bar.  We can improve that dialog, but maybe outside of this pull request.

> Checking for VDO support should be instant and done on page load. (Don't scan for the package via PackageKit when the dialog appears, in other words.)

I don't think we should do any PackageKit interactions on page load.  They are expensive and a bit unpredictable.

One goal here is to make people happy that don't care about VDO, don't have it installed, and will never install it.  If we would carry out expensive package installation simulations on every page load for stuff that they don't care about, we would unduly penalize them.

(In the bigger picture, I think we should just pre-install all the little utilities that a Cockpit page depends on and thus all this on-demand installation would only trigger on 'broken' systems.)

**Comment by mvollmer** [comment]
> I don't think we should do any PackageKit interactions on page load. They are expensive and a bit unpredictable.

And I realize that this is a good argument for clearly separating installation and creation into two steps. :)

**Comment by mvollmer** [comment]
@garrett: New video with the separate dialogs: https://www.youtube.com/watch?v=KDx0bp5IkpE

**Comment by andreasn** [comment]
Much better!
The install seems to take quite a while with little progress-reporting. Do you get any kind of data from the backend that can give you something like a rough estimation of a percentage of completion?

**Comment by mvollmer** [comment]
>  Is it possible to get any kind of percentage-estimation of the progress? What do you get from the backend?

We do get progress information from the background, but it is quite coarse.  In this case, I think you would maybe get five steps or so on the way to 100% and it would just sit at 60% or so for the whole time you see "Installng kmod-kvdo" in the video.

The progress would be smooth when there are tens or hundreds of packages, but not when there are only two, one of which takes a long time compiling a kernel module from source.

**Comment by mvollmer** [comment]
"vdo on demand" test needs adaptation.

**Comment by garrett** [comment]
>>    Checking for VDO support should be instant and done on page load. (Don't scan for the package via PackageKit when the dialog appears, in other words.)
>
> I don't think we should do any PackageKit interactions on page load. They are expensive and a bit unpredictable.

I strongly agree, that's why I specifically said to check for VDO support, but don't use PackageKit in that text you quoted. :wink: (Check to see if a VDO specific command exists, for example. If it does, it's installed. If it doesn't, it's not installed. Right?)

File: pkg/storaged/dialog.jsx [hunk 1]
```diff
@@ -275,7 +290,71 @@ function is_visible(field, values) {
...
-const Body = ({ body, fields, values, errors, onChange }) => {
+class Revealer extends React.Component {
+    constructor() {
+        super();
+        this.state = { revealed: false };
+    }
+
+    render() {
+        return (
+            <div>
```
Comment (garrett) [inline pkg/storaged/dialog.jsx]: Never use an empty <div> (or <span>).

This should either be classed or it should go away, replaced with a `<React.Fragment>` instead.

File: pkg/storaged/dialog.jsx [hunk 2]
```diff
@@ -275,7 +290,71 @@ function is_visible(field, values) {
...
+                <a onClick={event => { if (event.button == 0) this.setState({ revealed: !this.state.revealed }); }}>
+                    {this.props.summary}
+                </a> <i className={this.state.revealed ? "fa fa-angle-down" : "fa fa-angle-right"} />
```
Comment (garrett) [inline pkg/storaged/dialog.jsx]: Where's the text for accessibility? (Note: It should also be translatable.)

File: pkg/storaged/dialog.jsx [hunk 3]
```diff
@@ -275,7 +290,71 @@ function is_visible(field, values) {
...
+                <br />
```
Comment (garrett) [inline pkg/storaged/dialog.jsx]: Why do you have a `<br>`? They shouldn't exist in HTML code 99% of the time they're used (unless you really want a hard break — but you probably don't here).

File: pkg/storaged/dialog.jsx [hunk 4]
```diff
@@ -275,7 +290,71 @@ function is_visible(field, values) {
...
+    if (data.extra_names.length == 0)
+        summary = fmt_to_fragments(_("The $0 package will be installed."), <strong>{data.missing_names[0]}</strong>);
+    else
+        summary = fmt_to_fragments(cockpit.ngettext(_("The $0 package and $1 other package will be installed."),
+                                                    _("The $0 package and $1 other packages will be installed."),
+                                                    data.extra_names.length), <strong>{data.missing_names[0]}</strong>, data.extra_names.length);
+    if (data.remove_names.length > 0) {
+        summary = <>
+            {summary}
+            <br />
```
Comment (garrett) [inline pkg/storaged/dialog.jsx]: Why is there a `<br>`?

Either an element should be set to block or if it's text and you want paragraphs, you should use `<p>`s instead.

File: pkg/storaged/dialog.jsx [hunk 5]
```diff
@@ -275,7 +290,71 @@ function is_visible(field, values) {
...
+                                             data.remove_names.length),
+                            data.remove_names.length)}
+        </>;
+    }
+    if (data.extra_names.length > 0 || data.remove_names.length > 0) {
+        let extra_details = null;
+        let remove_details = null;
+        if (data.extra_names.length > 0)
+            extra_details = (
+                <div>
```
Comment (garrett) [inline pkg/storaged/dialog.jsx]: `<React.Fragment>`, please.

File: pkg/storaged/dialog.jsx [hunk 6]
```diff
@@ -275,7 +290,71 @@ function is_visible(field, values) {
...
+                    {_("Additional packages:")}
+                    <ul className="package-list-ct">{data.extra_names.map(id => <li key={id}>{id}</li>)}</ul>
+                </div>
+            );
+        if (data.remove_names.length > 0)
+            remove_details = (
+                <div>
```
Comment (garrett) [inline pkg/storaged/dialog.jsx]: Again, this should be a fragment, not a classless `<div>`.

File: pkg/storaged/dialog.jsx [hunk 7]
```diff
@@ -275,7 +290,71 @@ function is_visible(field, values) {
...
+                    {_("Removals:")}
+                    <ul className="package-list">{data.remove_names.map(id => <li key={id}>{id}</li>)}</ul>
+                </div>
+            );
+        summary = <><p>{summary}</p><Revealer summary={_("Details")}>{extra_details}{remove_details}</Revealer></>;
+    }
+
+    return <><br />{summary}</>;
```
Comment (garrett) [inline pkg/storaged/dialog.jsx]: Why the `<br>`?

File: pkg/storaged/storage-controls.jsx [hunk 8]
```diff
@@ -249,3 +250,29 @@ export class StorageUsageBar extends React.Component {
...
+        return <li><a onClick={this.props.onClick}>{this.props.children}</a></li>;
+    }
+}
+
+export class StorageBarMenu extends React.Component {
+    render() {
+        function toggle(excuse) {
+            const classes = "btn btn-primary" + (excuse ? " disabled" : "");
+            return (
+                <button className={classes} type="button" data-toggle="dropdown">
```
Comment (garrett) [inline pkg/storaged/storage-controls.jsx]: This has accessibility related issues. Please add text for screen readers.

File: pkg/storaged/storage.css [hunk 9]
```diff
@@ -723,3 +723,11 @@ td.storage-action {
 .sidepanel-row .spinner {
     display: inline-block;
 }
+
+.panel-heading {
+    position: static;
```
Comment (garrett) [inline pkg/storaged/storage.css]: Static is the default position for elements. Is this overriding something?

**Comment by mvollmer** [comment]
> > > Checking for VDO support should be instant and done on page load. (Don't scan for the package via PackageKit when the dialog appears, in other words.)
> > 
> > 
> > I don't think we should do any PackageKit interactions on page load. They are expensive and a bit unpredictable.
> 
> I strongly agree, that's why I specifically said to check for VDO support, but don't use PackageKit in that text you quoted. wink (Check to see if a VDO specific command exists, for example. If it does, it's installed. If it doesn't, it's not installed. Right?)

Ahh, I see.  Yes, we run `which insights-client` to figure out whether we need to bother with PackageKit.

When you said "Don't scan for the package via PackageKit when the dialog appears, in other words." I understood: Don't scan when the dialog opens, but rather scan on page load so that we don't need to anymore when the dialog opens.

File: pkg/storaged/storage.css [hunk 10]
```diff
@@ -723,3 +723,11 @@ td.storage-action {
...
+}
+
+.dropdown-toggle.btn-primary {
+    color: white;
```
Comment (garrett) [inline pkg/storaged/storage.css]: I'd prefer this to be uniquely classed in the HTML and then you use the class here, instead of telling it that all dropdown-toggles + primary buttons to have white text.

(Also, I'm pretty sure primary buttons have white text by default, so I'm not sure why this is here?)

**Comment by garrett** [comment]
I would be OKAY with this PR going in without the a11y fixes (but still with the React.Fragment and `<br>` fixes) provided that you do a follow-up PR that fixes the buttons in this PR and across the whole page, as the rest of the page has similar accessibility-related issues.

A lot of Cockpit has the same problem, but we need to fix it.

Details about icon issues (including icon buttons) @ https://github.com/cockpit-project/cockpit/issues/9325#issuecomment-550362240

**Comment by mvollmer** [comment]
> Please fix, and then I'll give the green light for this to be merged.

I have cheated and removed most of the code... The changes to the dialog code are not needed anymore.  I'll fix the rest.

File: pkg/storaged/storage.css (hunk 9 above)
Comment (mvollmer) [reply pkg/storaged/storage.css]: This overrides our own pk/lib/table.css:

```
/* Fix panel heading alignment & mobile layout */

.panel-heading {
    align-items: center;
    background: #f5f5f5;
    display: flex;
    flex-wrap: wrap;
    /* (28px small size widget height) + (0.5rem * 2) */
    min-height: calc(28px + 1rem);
    padding: 0.5rem 0.5rem 0.5rem 1rem;
    position: relative;
    z-index: 100;
}
```

The `position: relative` makes it so that the dropdown menu would show behind the "Drives" panel:

image

File: pkg/storaged/storage.css (hunk 10 above)
Comment (mvollmer) [reply pkg/storaged/storage.css]: This is indeed not necessary.  There was something that made it go black in the very beginning when focused, I think.  Maybe a bug in Patternfly that got fixed.

Comment (mvollmer) [reply pkg/storaged/storage.css]: Removed.

File: pkg/storaged/storage-controls.jsx (hunk 8 above)
Comment (mvollmer) [reply pkg/storaged/storage-controls.jsx]: I have added `aria-label="Create Devices"`.  Is that ok?

Comment (marusak) [reply pkg/storaged/storage-controls.jsx]: You can also translate it :) `aria-label={_("Create Devices")}` (although I am still not absolutely sure if we should translate those, but I guess why not?)

Comment (andreasn) [reply pkg/storaged/storage-controls.jsx]: I think @Gundersanne asked if we wanted to do that for the Firewall accessibility fixes. I think we landed on Yes, since you want to be able to use Cockpit if you're both visually impaired and don't speak English.

File: pkg/storaged/storage.css (hunk 9 above)
Comment (garrett) [reply pkg/storaged/storage.css]: Oh! Well then.

I assume the `position: relative` is because we're using floats* and want to make sure they don't escape.

_* **ugh!** Floats shouldn't be used for layout. I've gotten rid of a bunch, but we're still using a lot of floats and `pull-right`s (which use floats)._

Why shouldn't floats be used for layout? It's a hack. They're fragile. Also, in order to use them, you have to mess with document order (put the rightmost thing first) and this breaks accessibility (among other things).

But meanwhile, we do have some lingering floats and they need to be contained, so it needs to still be there for now, I guess.

_**TL;DR**: OK, you make a very good point and we do need `position: static` here. :wink:_

**Comment by martinpitt** [comment]
Before reviewing the code, I tested this a bit. I have a standard Fedora LVM install. On master:

storage-devs-master

Apparently we never treated LVs as a specially recognized thing, just as "some random block device"? But now with the merged view this looks a bit odd:

storage-devs-merged

Wouldn't it make sense to merge "Devices" and "Other Devices" into one box? There is no clear difference between these any more.

When I create a RAID device (out of  two partitions on scsi_debug), the layout also seems to stay compatible to what "Other Devices" shows:

storage-devs-merged-raid

This is fine for a follow-up (I'm happy to put this into an issue), but I'd like to at least discuss it quickly before we land this.

> I have added "Create Loop Device",

I don't see this, on my system the only menu entry is "Create RAID device". Was that not implemented, or am I missing some package?

Otherwise this works well on my system (rather minimalist F31 OSTree). I'll test it some more in a VM.

**Comment by mvollmer** [comment]
> > I have added "Create Loop Device",
> 
> I don't see this, on my system the only menu entry is "Create RAID device". Was that not implemented, or am I missing some package?

This was not implemented, after all.

**Comment by mvollmer** [comment]
> Apparently we never treated LVs as a specially recognized thing, just as "some random block device"?

No, you should have "fedora" Volume Group entry in "Devices", and no "Other Devices" at all.  "Other Devices" should only show the block devices that we don't specifically recognize, like loop back devices.

You need to have `udisks2-lvm2` installed.

File: pkg/storaged/overview.jsx [hunk 11]
```diff
@@ -24,9 +24,7 @@ import { StoragePlots } from "./plot.jsx";
 
 import { FilesystemsPanel } from "./fsys-panel.jsx";
 import { NFSPanel } from "./nfs-panel.jsx";
-import { MDRaidsPanel } from "./mdraids-panel.jsx";
-import { VGroupsPanel } from "./vgroups-panel.jsx";
-import { VDOsPanel } from "./vdos-panel.jsx";
+import { ThingsPanel } from "./things-panel.jsx";
```
Comment (martinpitt) [inline pkg/storaged/overview.jsx]: StorageDevicesPanel wouldn't have hurt either, but let's let that slip as developer's humor :grin:
